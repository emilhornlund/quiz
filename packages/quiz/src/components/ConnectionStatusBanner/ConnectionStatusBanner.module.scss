@import '../../styles/colors';
@import '../../styles/helpers';

.connectionStatusBanner {
  display: flex;
  flex-direction: row;
  justify-content: center;
  position: absolute;
  left: 0;
  right: 0;

  @include mobile {
    top: $mobile-layout-gutter;
  }

  @include tablet {
    top: $tablet-layout-gutter;
  }

  @include desktop {
    top: $desktop-layout-gutter;
  }

  .banner {
    display: flex;
    flex-direction: column;
    box-sizing: border-box;
    color: $white;
    font-weight: bold;
    appearance: none;
    outline: none;
    border: none;

    @include mobile {
      padding: $mobile-layout-gutter * 0.5 $mobile-layout-gutter;
      border-radius: $layout-gutter * 0.5;

      @include fontSize($mobile-text-font-size);
    }

    @include tablet {
      padding: $tablet-layout-gutter * 0.5 $tablet-layout-gutter;
      border-radius: $layout-gutter * 0.75;

      @include fontSize($tablet-text-font-size);
    }

    @include desktop {
      padding: $desktop-layout-gutter * 0.5 $desktop-layout-gutter;
      border-radius: $layout-gutter;

      @include fontSize($desktop-text-font-size);
    }

    &.success {
      background-color: $green-2;
    }

    &.warning {
      background-color: $orange-2;
    }

    &.error {
      background-color: $red-2;
    }
  }
}
