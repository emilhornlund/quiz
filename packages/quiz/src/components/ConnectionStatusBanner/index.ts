export type { ConnectionStatusBannerProps } from './ConnectionStatusBanner'
export { default } from './ConnectionStatusBanner'
