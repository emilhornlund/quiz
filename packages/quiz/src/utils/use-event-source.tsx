import { GameEvent } from '@quiz/common'
import { useCallback, useEffect, useRef, useState } from 'react'

export enum ConnectionStatus {
  CONNECTED = 'CONNECTED',
  RECONNECTING = 'RECONNECTING',
  RECONNECTING_FAILED = 'RECONNECTING_FAILED',
}

export const useEventSource = (
  gameID?: string,
): [GameEvent | undefined, ConnectionStatus] => {
  const eventSourceRef = useRef<EventSource>()

  const [gameEvent, setGameEvent] = useState<GameEvent>()

  const [connectionStatus, setConnectionStatus] = useState<ConnectionStatus>(
    ConnectionStatus.CONNECTED,
  )

  const MAX_RETRIES = 10

  const createEventSource = useCallback((gameID: string, retryCount = 0) => {
    if (retryCount >= MAX_RETRIES) {
      console.error(
        'Max retry attempts reached. Stopping reconnection attempts.',
      )
      setConnectionStatus(ConnectionStatus.RECONNECTING_FAILED)
      return
    }

    const eventSource = new EventSource(
      `http://localhost:3000/api/game/${gameID}/events`,
    )

    eventSourceRef.current = eventSource

    eventSource.onmessage = (event) => {
      setGameEvent(JSON.parse(event.data) as GameEvent)
      retryCount = 0
      setConnectionStatus(ConnectionStatus.CONNECTED)
    }

    eventSource.onerror = () => {
      setConnectionStatus(ConnectionStatus.RECONNECTING)
      eventSource.close()

      const nextRetryDelay = Math.min(1000 * 2 ** retryCount, 30000)
      console.error(
        `Connection failed, retrying in ${nextRetryDelay / 1000} seconds...`,
      )

      setTimeout(() => {
        createEventSource(gameID, retryCount + 1)
      }, nextRetryDelay)
    }
  }, [])

  useEffect(() => {
    if (gameID) {
      if (eventSourceRef.current) {
        eventSourceRef.current.close()
      }

      createEventSource(gameID)

      return () => {
        eventSourceRef.current?.close()
      }
    }
  }, [createEventSource, gameID])

  return [gameEvent, connectionStatus]
}
