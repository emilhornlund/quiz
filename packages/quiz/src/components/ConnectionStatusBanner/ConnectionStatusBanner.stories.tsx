import type { Meta, StoryObj } from '@storybook/react'

import { ConnectionStatus } from '../../utils/use-event-source.tsx'

import ConnectionStatusBanner from './ConnectionStatusBanner'

const meta = {
  component: ConnectionStatusBanner,
} satisfies Meta<typeof ConnectionStatusBanner>

export default meta
type Story = StoryObj<typeof meta>

export const Connected = {
  args: {
    status: ConnectionStatus.CONNECTED,
  },
} satisfies Story

export const Reconnecting = {
  args: {
    status: ConnectionStatus.RECONNECTING,
  },
} satisfies Story

export const ReconnectingFailed = {
  args: {
    status: ConnectionStatus.RECONNECTING_FAILED,
  },
} satisfies Story
