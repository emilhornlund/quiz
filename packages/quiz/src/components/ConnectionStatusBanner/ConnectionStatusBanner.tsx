import React from 'react'

import { classNames } from '../../utils/helpers.ts'
import { ConnectionStatus } from '../../utils/use-event-source.tsx'

import styles from './ConnectionStatusBanner.module.scss'

const classNameFromStatus: { [key in ConnectionStatus]: string } = {
  [ConnectionStatus.CONNECTED]: styles.success,
  [ConnectionStatus.RECONNECTING]: styles.warning,
  [ConnectionStatus.RECONNECTING_FAILED]: styles.error,
}

const messageFromStatus: { [key in ConnectionStatus]: string } = {
  [ConnectionStatus.CONNECTED]: 'Connected',
  [ConnectionStatus.RECONNECTING]: 'Reconnecting...',
  [ConnectionStatus.RECONNECTING_FAILED]: 'Reconnecting failed...',
}

export interface ConnectionStatusBannerProps {
  status: ConnectionStatus
}

const ConnectionStatusBanner: React.FC<ConnectionStatusBannerProps> = ({
  status,
}) => {
  const [show, setShow] = React.useState<boolean>(true)

  return (
    <div className={styles.connectionStatusBanner}>
      {show && (
        <button
          id="status"
          name="status"
          type="button"
          className={classNames(styles.banner, classNameFromStatus[status])}
          onClick={() => setShow(false)}>
          {messageFromStatus[status]}
        </button>
      )}
    </div>
  )
}

export default ConnectionStatusBanner
