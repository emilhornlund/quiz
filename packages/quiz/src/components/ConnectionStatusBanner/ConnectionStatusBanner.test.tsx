import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import { ConnectionStatus } from '../../utils/use-event-source.tsx'

import ConnectionStatusBanner from './ConnectionStatusBanner'

describe('ConnectionStatusBanner', () => {
  it('should render ConnectionStatusBanner with connected status', () => {
    const { container } = render(
      <ConnectionStatusBanner status={ConnectionStatus.CONNECTED} />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render ConnectionStatusBanner with reconnecting status', () => {
    const { container } = render(
      <ConnectionStatusBanner status={ConnectionStatus.RECONNECTING} />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render ConnectionStatusBanner with reconnecting failed status', () => {
    const { container } = render(
      <ConnectionStatusBanner status={ConnectionStatus.RECONNECTING_FAILED} />,
    )

    expect(container).toMatchSnapshot()
  })
})
