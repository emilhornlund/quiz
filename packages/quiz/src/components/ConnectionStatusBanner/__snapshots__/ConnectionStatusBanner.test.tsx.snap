// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectionStatusBanner > should render ConnectionStatusBanner with connected status 1`] = `
<div>
  <div
    class="connectionStatusBanner"
  >
    <button
      class="banner success"
      id="status"
      name="status"
      type="button"
    >
      Connected
    </button>
  </div>
</div>
`;

exports[`ConnectionStatusBanner > should render ConnectionStatusBanner with reconnecting failed status 1`] = `
<div>
  <div
    class="connectionStatusBanner"
  >
    <button
      class="banner error"
      id="status"
      name="status"
      type="button"
    >
      Reconnecting failed...
    </button>
  </div>
</div>
`;

exports[`ConnectionStatusBanner > should render ConnectionStatusBanner with reconnecting status 1`] = `
<div>
  <div
    class="connectionStatusBanner"
  >
    <button
      class="banner warning"
      id="status"
      name="status"
      type="button"
    >
      Reconnecting...
    </button>
  </div>
</div>
`;
